cmd/kubeadm/app/apis/kubeadm/types.go

```diff
@@ -434,6 +434,11 @@ type ComponentConfig interface {
 
 	// Default patches the component config with kubeadm preferred defaults
 	Default(cfg *ClusterConfiguration, localAPIEndpoint *APIEndpoint)
+
+	// UserSupplied returns true if the component config was loaded from file or a config map
+	// or false if it was generated by kubeadm and has no user values in it.
+	// This call is not applicable to addon component configs. The behavior in such cases is undefined.
```

Comment:
Sorry but I don't' understand this point. Why this does not apply to kube-proxy?

Reply:
So this basically is the alternative implementation of #86070. The basic idea here is to treat the presence of a config map as an indication that a component config was supplied by an user. If there is no config map, then the config was kubeadm generated and can be safely regenerated.
The reason as to why this does not apply to kube-proxy is that kube-proxy is deployed as a daemon set and fetches its config from a config map. Hence the config map is always present and treated as user supplied (requiring a manual conversion if its version changes).

For the protocol this is the version that I don't like. I think that the checksum implementation (the one in #86070) is safer, better, and extends to kube-proxy too (kubeadm generated kube-proxy config does not have to be manually upgraded on a config version change).